File: Cpp/889_Construct_Binary_Tree_from_Preorder_and_Postorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructFromPrePost(vector<int>& pre, vector<int>& post) {
        stack<TreeNode*> parent;
        
        TreeNode * root = new TreeNode(pre[0]);
        TreeNode * current = root;
        
        parent.push(current);
        
        int post_index = 0;
        for (int pre_index = 1; pre_index < pre.size();) {
            auto found = find(pre.begin(), pre.begin()+pre_index, post[post_index]);
            if (found != pre.begin()+pre_index){
                current = parent.top();
                parent.pop();
                ++post_index;
            } else {
                parent.push(current);
                TreeNode * child = new TreeNode(pre[pre_index]);
                
                if (!current->left) {
                    current->left = child;
                } else if(!current->right) {
                    current->right = child;
                } 
                current = child;
                ++pre_index;
            }
        }
        return root;
    }
};

File: Kotlin/889_Construct_Binary_Tree_from_Preorder_and_Postorder_Traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun constructFromPrePost(pre: IntArray, post: IntArray): TreeNode? {
        var stack = Stack<TreeNode>()
        
        var root = TreeNode(pre[0])
        var current = root
        
        stack.push(current)
        
        var pre_index = 1
        var post_index = 0
        while(pre_index < pre.size){
            var found = pre.indexOfFirst({it -> it == post[post_index]})
            
            if(found < pre_index){
                current = stack.pop()
                ++post_index
            } else {
                stack.push(current)
                var child = TreeNode(pre[pre_index])
                
                if(current.left == null) {
                    current.left = child
                } else if(current.right == null) {
                    current.right = child
                }
                current = child
                ++pre_index
            }
        }
        
        return root
    }
}